README.md
# lighter-run
[![Chat](https://badges.gitter.im/chat.svg)](//gitter.im/lighterio/public)
[![Version](https://img.shields.io/npm/v/lighter-run.svg)](//www.npmjs.com/package/lighter-run)
[![Downloads](https://img.shields.io/npm/dm/lighter-run.svg)](//www.npmjs.com/package/lighter-run)
[![Build](https://img.shields.io/travis/lighterio/lighter-run.svg)](//travis-ci.org/lighterio/lighter-run)
[![Coverage](https://img.shields.io/codecov/c/github/lighterio/lighter-run/master.svg)](//codecov.io/gh/lighterio/lighter-run)
[![Style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](//www.npmjs.com/package/standard)

With `lighter-run`, you can run the node application that's in your current
directory, and it will restart when files change.

## Quick Start
1. Install globally (using sudo if necessary).
```bash
npm install --global lighter-run
```
2. Ensure that the `"main"` property in your `package.json` points to your
application's entry point file.

3. Run!
```bash
lighter-run
```

## Configuration
`lighter-run` uses `lighter-config` for its configuration. Just create a
`"config/base.json"` file in your project, and add some options under a
property called `"lighterRun"`. The following is an example of a configuration
which uses the default values:
```json
{
  "lighterRun": {
    "minRestartDelay": 500,
    "maxRestartDelay": 5000,
    "restartDelayBackoff": 2,
    "cleanTime": 2000,
    "ignore": [
      ".DS_Store",
      ".cache",
      ".git",
      ".idea",
      ".project",
      "coverage",
      "data",
      "log"
    ],
    "live": [
      "public",
      "scripts",
      "styles",
      "views"
    ],
    "watchDirs": null
  }
}
```

**minRestartDelay**<br>
Amount of time to wait before trying to restart (the first time).

**maxRestartDelay**<br>
Maximum amount of time to wait before trying to restart, after failing many times.

**restartDelayBackoff**<br>
Multiplier to be applied to the restart delay time after each failure.

**cleanTime**<br>
Length of time that a process must be running before it's considered to have started cleanly.

**ignore**<br>
File patterns in which changes should be ignored.

**live**<br>
File patterns that can be live-reloaded instead of restarting the process.

**watchDirs**<br>
Absolute paths for directories that should be watched. (Defaults to the current
working directory).


## More on lighter-run...
* [Contributing](//github.com/lighterio/lighter-run/blob/master/CONTRIBUTING.md)
* [License (ISC)](//github.com/lighterio/lighter-run/blob/master/LICENSE.md)
* [Change Log](//github.com/lighterio/lighter-run/blob/master/CHANGELOG.md)
* [Roadmap](//github.com/lighterio/lighter-run/blob/master/ROADMAP.md)


CHANGELOG.md
# lighter-run Change Log

## v1.2.0 - 2017-03-06
* `minor` Added an optional file argument.

## v1.1.0 - 2017-01-05
* `minor` Added **watchDirs** and **ignore** config properties.

## v1.0.0 - 2016-10-25
* `major` First release.


test/run.unit.js
'use strict'
/* global describe it is mock unmock */

var cp = require('child_process')
var cwd = process.cwd()
var cache = require.cache
var runJs = cwd + '/run.js'
var fseventsJs = cwd + '/node_modules/fsevents/fsevents.js'
var noop = function () {}
var fakeStream = {pipe: noop}

describe('run', function () {
  it('spawns a child process', function (done) {
    mock(cp, {
      spawn: function (node, args) {
        is.string(node)
        is.array(args)
        unmock(cp)
        setTimeout(done, 1)
        return {
          on: noop,
          stdout: fakeStream,
          stderr: fakeStream
        }
      }
    })
    delete cache[runJs]
    require('../run')
  })

  it('logs an error if there is no package.json', function (done) {
    mock(console, {
      error: mock.count()
    })
    mock(process, {
      exit: function () {
        is(console.error.value, 1)
        unmock(console)
        unmock(process)
        process.chdir(cwd)
        done()
      }
    })
    process.chdir(__dirname)
    delete cache[runJs]
    require('../run')
  })

  it('listens for changes', function (done) {
    var fsevents = cache[fseventsJs]
    var changed
    mock(console, {
      log: mock.args()
    })
    mock(cp, {
      spawn: function (node, args) {
        return {
          on: noop,
          stdout: fakeStream,
          stderr: fakeStream,
          kill: function () {
            unmock(cp)
            unmock(fsevents)
            unmock(console)
            done()
          }
        }
      }
    })
    mock(fsevents, {
      exports: function () {
        return {
          on: function (type, fn) {
            changed = fn
          },
          start: mock.count()
        }
      }
    })
    delete cache[runJs]
    require('../run')
    changed(__filename, {
      path: __filename,
      event: 'modified'
    })
  })
})


test/spawn.unit.js
'use strict'
/* global describe it is mock unmock beforeEach afterEach */

var config = require('lighter-config')
var cp = require('child_process')
var Child = require('./helpers/child.helper.js')
var cwd = process.cwd()
var stdout = process.stdout
var write = stdout.write
var cache = require.cache
var runJs = cwd + '/run.js'

describe('spawn', function () {
  beforeEach(function () {
    mock(config, {
      lighterRun: {
        minRestartDelay: 1,
        maxRestartDelay: 4,
        restartDelayBackoff: 2,
        cleanTime: 20
      }
    })
    mock(stdout, {
      write: function () {}
    })
  })

  afterEach(function () {
    unmock(config)
    unmock(cp)
    unmock(process)
    stdout.write = write
  })

  it('spawns a child process', function (done) {
    mock(cp, {
      spawn: function (node, args) {
        is.string(node)
        is.array(args)
        setTimeout(done, 1)
        return new Child()
      }
    })
    delete cache[runJs]
    require('../run')
  })

  it('restarts the child process when it dies', function (done) {
    var spawnCount = 0
    mock(cp, {
      spawn: function (node, args) {
        var child = new Child()
        if (spawnCount++) {
          done()
        } else {
          setTimeout(function () {
            child.close()
          }, 1)
        }
        return child
      }
    })
    delete cache[runJs]
    require('../run')
  })

  it('writes output once the new child process runs cleanly', function (done) {
    var spawnCount = 0
    var child = new Child()
    mock(stdout, {
      write: mock.args()
    })
    mock(cp, {
      spawn: function (node, args) {
        ++spawnCount
        if (spawnCount < 4) {
          setTimeout(function () {
            stdout.write('Spawn ' + spawnCount + '.')
            child.close()
          }, spawnCount * 10)
        } else {
          setTimeout(done, 1)
        }
        child = new Child()
        return child
      }
    })
    delete cache[runJs]
    require('../run')
  })

  it('skips the dot if the child was killed manually', function (done) {
    var spawnCount = 0
    var child = new Child()
    mock(stdout, {
      write: mock.args()
    })
    mock(cp, {
      spawn: function (node, args) {
        ++spawnCount
        if (spawnCount < 4) {
          setTimeout(function () {
            stdout.write('Spawn ' + spawnCount + '.')
            child.killed = true
            child.close()
          }, spawnCount * 10)
        } else {
          is.same(stdout.write.value, [
            {'0': 'Spawn 1.'},
            {'0': ''},
            {'0': 'Spawn 3.'},
            {'0': '\n'}
          ])
          setTimeout(done, 1)
        }
        child = new Child()
        return child
      }
    })
    delete cache[runJs]
    require('../run')
  })
})


test/watcher.unit.js
'use strict'
/* global describe it is mock unmock */

if (process.platform === 'darwin') {
  require('fsevents')
  var cp = require('child_process')
  var cwd = process.cwd()
  var cache = require.cache
  var runJs = cwd + '/run.js'
  var fseventsJs = cwd + '/node_modules/fsevents/fsevents.js'
  var noop = function () {}
  var fakeStream = {
    pipe: noop,
    on: noop,
    emit: noop
  }

  describe('watcher', function () {
    it('listens for changes and kills the child process', function (done) {
      var fsevents = cache[fseventsJs]
      var changed
      mock(console, {
        log: mock.args()
      })
      mock(cp, {
        spawn: function (node, args) {
          return {
            on: noop,
            stdout: fakeStream,
            stderr: fakeStream,
            kill: function () {
              is.in(console.log.value[0][0], 'Deleted "/tmp/test.txt"')
              unmock(console)
              unmock(cp)
              unmock(fsevents)
              done()
            }
          }
        }
      })
      mock(fsevents, {
        exports: function () {
          return {
            on: function (type, fn) {
              changed = fn
            },
            start: noop
          }
        }
      })
      delete cache[runJs]
      require('../run')
      changed('/tmp/test.txt', {
        path: '/tmp/test.txt',
        event: 'deleted'
      })
    })

    it('removes the current directory from the file path', function (done) {
      var fsevents = cache[fseventsJs]
      var changed
      mock(console, {
        log: mock.args()
      })
      mock(cp, {
        spawn: function (node, args) {
          return {
            on: noop,
            stdout: fakeStream,
            stderr: fakeStream,
            kill: function () {
              is.in(console.log.value[0][0], 'Modified "test/watcher.unit.js"')
              unmock(console)
              unmock(cp)
              unmock(fsevents)
              done()
            }
          }
        }
      })
      mock(fsevents, {
        exports: function () {
          return {
            on: function (type, fn) {
              changed = fn
            },
            start: mock.count()
          }
        }
      })
      delete cache[runJs]
      require('../run')
      changed(__filename, {
        path: __filename,
        event: 'modified'
      })
    })

    it('only kills the child process once', function (done) {
      var fsevents = cache[fseventsJs]
      var killCount = 0
      var child = {
        on: noop,
        stdout: fakeStream,
        stderr: fakeStream,
        kill: function () {
          killCount++
          this.killed = true
        }
      }
      var changed
      mock(console, {
        log: mock.count()
      })
      mock(cp, {
        spawn: function (node, args) {
          return child
        }
      })
      mock(fsevents, {
        exports: function () {
          return {
            on: function (type, fn) {
              changed = fn
            },
            start: mock.count()
          }
        }
      })
      delete cache[runJs]
      require('../run')
      changed(__filename, {
        path: __filename,
        event: 'modified'
      })
      changed(__filename, {
        path: __filename,
        event: 'deleted'
      })
      setTimeout(function () {
        is(console.log.value, 2)
        is(killCount, 1)
        unmock(console)
        unmock(cp)
        unmock(fsevents)
        done()
      }, 1)
    })

    it('writes live-reloadable paths to child.stdin instead of killing the child process', function (done) {
      var fsevents = cache[fseventsJs]
      var changed
      mock(console, {
        log: noop
      })
      mock(cp, {
        spawn: function (node, args) {
          return {
            on: noop,
            stdout: fakeStream,
            stderr: fakeStream,
            stdin: {
              write: function (output) {
                is.string(output)
                unmock(console)
                unmock(cp)
                unmock(fsevents)
                done()
              }
            }
          }
        }
      })
      mock(fsevents, {
        exports: function () {
          return {
            on: function (type, fn) {
              changed = fn
            },
            start: noop
          }
        }
      })
      delete cache[runJs]
      require('../run')
      changed(cwd + '/.cache/run.json', {
        path: cwd + '/.cache/run.json',
        event: 'created'
      })
    })

    it('allows live-reload configuration via config.lighterRun.live', function (done) {
      var fsevents = cache[fseventsJs]
      var config = require('lighter-config')
      var changed
      mock(config, {
        lighterRun: {
          live: '/*.log'
        }
      })
      mock(console, {
        log: noop
      })
      mock(cp, {
        spawn: function (node, args) {
          return {
            on: noop,
            stdout: fakeStream,
            stderr: fakeStream,
            stdin: {
              write: function (output) {
                is.string(output)
                unmock(config)
                unmock(console)
                unmock(cp)
                unmock(fsevents)
                done()
              }
            }
          }
        }
      })
      mock(fsevents, {
        exports: function () {
          return {
            on: function (type, fn) {
              changed = fn
            },
            start: noop
          }
        }
      })
      delete cache[runJs]
      require('../run')
      changed(cwd + '/run.log', {
        path: cwd + '/run.log',
        event: 'modified'
      })
    })

    it('falls back to fs.watch', function (done) {
      var fs = require('fs')
      var fsevents = cache[fseventsJs]
      mock(console, {
        log: mock.count()
      })
      mock(cp, {
        spawn: function () {
          return {
            on: noop,
            stdout: fakeStream,
            stderr: fakeStream,
            stdin: fakeStream,
            kill: function () {
              is(console.log.value, 1)
              unmock(console)
              unmock(cp)
              unmock(fsevents)
              unmock(fs)
              done()
            }
          }
        }
      })
      mock(fsevents, {
        exports: function () {
          throw new Error('fsevents only works on OSX.')
        }
      })
      mock(fs, {
        watch: function (dir, fn) {
          is(dir, cwd)
          setTimeout(function () {
            fn('change', cwd + '/restart.js')
          }, 1)
        }
      })
      delete cache[runJs]
      require('../run')
    })
  })
}


run.js
#!/usr/bin/env node

var spawn = require('child_process').spawn
var config = require('lighter-config').lighterRun || {}
var stdout = process.stdout
var write = stdout.write
var node = process.execPath
var argv = process.argv
var cwd = process.cwd()
var env = process.env
var child = null
var output = ''
var previousStart = 0
var failureOutput

// Try to restart in half a second.
var minRestartDelay = config.minRestartDelay || 500

// Wait 5 seconds at most.
var maxRestartDelay = config.maxRestartDelay || 5000

// Wait twice as long each time.
var restartDelayBackoff = config.restartDelayBackoff || 2

// Call a restart "ok" after 2 seconds without failing.
var cleanTime = config.cleanTime || 2000

// The first time we restart, do it quickly.
var restartDelay = 0

// Directories to watch.
var watchDirs = config.watchDirs || [cwd]

// Globs for paths whose changes should be ignored.
var ignore = globify(config.ignore || ['.DS_Store', '.cache', '.git', '.idea', '.project', 'coverage', 'data', 'log'])

// Globs for paths whose changes are live-reloadable.
var live = globify(config.live || ['public', 'scripts', 'styles', 'views'])

/**
 * Treat a single string or a falsy value as a string array.
 *
 * @param  {String|Array} value  A string or an array of strings.
 * @return {Array}               An array of RegExp patterns.
 */
function globify (value) {
  var array = typeof value === 'string' ? [value] : value
  array.forEach(function (pattern, index) {
    pattern = pattern
      .replace(/([\W])/g, function (c) {
        switch (c) {
          case '*': return '.+'
          default: return '\\' + c
        }
      })
    if (pattern.substr(0, 2) === '\\\/') {
      pattern = '^' + pattern.substr(2)
    } else {
      pattern = '(^|\\/)' + pattern
    }
    array[index] = pattern + '($|\\/)'
  })
  var pattern = '(' + array.join('|') + ')'
  pattern = new RegExp(pattern, 'i')
  return pattern
}

// Watch for changes.
watchDirs.forEach(function (dir) {
  var fsevents
  try {
    fsevents = require('fsevents')
    var watcher = fsevents(dir)
    watcher.on('change', changed)
    watcher.start()
  } catch (e) {
    var fs = require('fs')
    fs.watch(dir, function (type, path) {
      changed(path, {event: type, path: path})
    })
  }
})

// Find arguments that come before or after a double-dash.
var dashes = argv.indexOf('--')
var args = dashes < 0 ? argv : argv.slice(0, dashes)
var childArgs = dashes < 0 ? [] : argv.slice(dashes + 1)

// Find the specified file, or default to "main" from "package.json".
try {
  var file = args[2] || require.resolve(cwd)
  childArgs.unshift(file)
} catch (e) {
  console.error('The current directory does not have a node application.\n' +
    'Please use "npm init", then create an entry point file such as "index.js".')
  process.exit()
}

// Start the application!
start()

/**
 * Respond to a change event.
 *
 * @param  {String} path  Path of the file in which the change occurred.
 * @param  {Object} info  Information about the change from `fsevents`.
 */
function changed (path, info) {
  // Get a relative path.
  if (path.indexOf(cwd + '/') === 0) {
    path = path.slice(cwd.length + 1)
  }
  // If not ignored, we'll at least log the change.
  if (!ignore.test(path)) {
    var data = info.event
    data = data[0].toUpperCase() + data.substr(1) + ' "' + path +
      '"\u001b[90m at ' + (new Date()).toTimeString() + '\u001b[39m'
    if (!child.killed) {
      if (live.test(path)) {
        console.log('\u001b[32m' + data)
        info = JSON.stringify(info)
        child.stdin.write(info)
      } else {
        console.log('\u001b[33m' + data + '\n')
        child.kill()
      }
    }
  }
}

/**
 * Get a string of numberless ordered lines for deduping logs.
 *
 * @param  {String} text  Text received from the child's stdout.
 * @return {String}       Numberless ordered lines.
 */
function munge (text) {
  var lines = ('' + text).split('\n')
  lines.forEach(function (line, index) {
    lines[index] = lines[index].replace(/\d+/, '#')
  })
  lines.sort()
  text = lines.join('\n').trim()
  return text
}

/**
 * Start an application.
 */
function start () {
  var now = Date.now()
  var elapsed = now - previousStart

  // If it's been a while since we restarted, call this a clean start.
  var isCleanStart = elapsed >= cleanTime
  previousStart = now

  // Spawn the child process, and pipe output to stdout.
  child = spawn(node, childArgs, {cwd: cwd, env: env})
  child.stdout.pipe(stdout)
  child.stderr.pipe(stdout)

  // When we've started cleanly, pipe child process output directly to stdout.
  if (isCleanStart) {
    restartDelay = 0

  // After a fast failure, buffer the output in case we fail again.
  } else {
    stdout.write = function (chunk) {
      output += chunk
    }

    // When we've started cleanly, write output to stdout and start piping.
    this.cleanTimer = setTimeout(function () {
      stdout.write = write
      stdout.write(output)
      output = ''
      restartDelay = 0
    }, cleanTime)
  }

  // When a child process dies, restart it.
  child.on('close', function () {
    // Restore stdout.write
    stdout.write = write

    // If we failed differently, log the new output.
    if (failureOutput && (munge(output) !== munge(failureOutput))) {
      stdout.write('\n' + output)

    // If we failed the same way, just show another red dot.
    } else if (!child.killed) {
      stdout.write('\u001b[31m.\u001b[39m')
    }
    failureOutput = output
    output = ''
    clearTimeout(this.cleanTimer)
    this.cleanTimer = setTimeout(start, restartDelay)
    restartDelay = Math.min(restartDelay * restartDelayBackoff, maxRestartDelay) || minRestartDelay
  })
}
